src/Utilities/TaggedTuple.hpp

```diff
@@ -705,13 +705,61 @@ template <template <typename...> class List, typename... Tags>
 struct tagged_tuple_typelist_impl<List<Tags...>> {
   using type = TaggedTuple<Tags...>;
 };
+
+template <typename T>
+struct typelist_from_tagged_tuple_impl;
+
+template <typename... Tags>
+struct typelist_from_tagged_tuple_impl<TaggedTuple<Tags...>> {
+  using type = tmpl::list<Tags...>;
+};
 }  // namespace TaggedTuple_detail
 
 /// \ingroup UtilitiesGroup
 template <typename T>
 using tagged_tuple_from_typelist =
     typename TaggedTuple_detail::tagged_tuple_typelist_impl<T>::type;
 
+template <typename T>
+using typelist_from_tagged_tuple =
+    typename TaggedTuple_detail::typelist_from_tagged_tuple_impl<T>::type;
+
+namespace TaggedTuple_detail {
+
+template <typename T>
+struct reorder_tagged_tuple_impl;
+
+template <typename... Tags>
+struct reorder_tagged_tuple_impl<TaggedTuple<Tags...>> {
+  template <typename InputTaggedTuple>
+  static TaggedTuple<Tags...> apply(InputTaggedTuple&& input) noexcept {
+    static_assert(
+        cpp17::is_same_v<
+            tmpl::list_difference<tmpl::list<Tags...>,
+                                  typelist_from_tagged_tuple<InputTaggedTuple>>,
+            tmpl::list<>> and
+            cpp17::is_same_v<tmpl::list_difference<
+                                 typelist_from_tagged_tuple<InputTaggedTuple>,
+                                 tmpl::list<Tags...>>,
+                             tmpl::list<>>,
+        "The input and output TaggedTuples must be the same except"
+        "for ordering.");
+    return TaggedTuple<Tags...>(std::move(get<Tags>(input))...);
```

Comment:
You have a bug here. Since you are using forwarding references and them `move`ing the members of `input` you can destroy `input` if it is passed in by a non-const l-value reference. I think there should be 2 implementations, one which takes a const l-value `TaggedTuple` and one which takes a non-const r-value `TaggedTuple`. You can then also get rid of the `typelist_from_tagged_tuple` since you'll have the parameter pack. Specifically,

```cpp
template <typename... InputTags>
static TaggedTuple<Tags...> apply(TaggedTuple<InputTags>&& input) noexcept {
  static_assert(...);
  return TaggedTuple<Tags...>(std::move(get<Tags>(input))...);
}

template <typename... InputTags>
static TaggedTuple<Tags...> apply(const TaggedTuple<InputTags>& input) noexcept {
  static_assert(...);
  return TaggedTuple<Tags...>(get<Tags>(input)...);
}
```